File: src/offer/Questions25.java
package offer;

import offer.LinkedList.Node;

//合并两个已排序的链表

public class Questions25 {
	
	public static void main(String[] arg0) {
		Node<Integer> node1 = new Node<>();
		node1.set(1);
		Node<Integer> node2 = new Node<>();
		node2.set(3);
		node1.setNext(node2);
		Node<Integer> node3 = new Node<>();
		node3.set(5);
		node2.setNext(node3);
		Node<Integer> node4 = new Node<>();
		node4.set(2);
		Node<Integer> node5 = new Node<>();
		node5.set(4);
		node4.setNext(node5);
		Node<Integer> node6 = new Node<>();
		node6.set(6);
		node5.setNext(node6);
		//Node<Integer> node = merge1(node1, node4);
		Node<Integer> node = merge2(node1, node4);
		while (node != null) {
			System.out.println(node.get());
			node = node.getNext();
		}
	}
	
	//使用循环
	public static Node<Integer> merge1(Node<Integer> node1, Node<Integer> node2) {
		if (node1 == null && node2 == null) {
			throw new IllegalArgumentException("输入的链表不能为空");
		}
		if (node1 == null) {
			return node2;
		}
		if (node2 == null) {
			return node1;
		}
		Node<Integer> head;
		Node<Integer> tail;
		if (node1.get() >= node2.get()) {
			head = tail = node2;
			node2 = node2.getNext();
		} else {
			head = tail = node1;
			node1 = node1.getNext();
		}
		while (node1 != null && node2 != null) {
			if (node1.get() >= node2.get()) {
				tail.setNext(node2);
				tail = node2;
				node2 = node2.getNext();
			} else {
				tail.setNext(node1);
				tail = node1;
				node1 = node1.getNext();
			}
		}
		if (node1 == null) {
			while (node2 != null) {
				tail.setNext(node2);
				node2 = node2.getNext();
			}
		}
		if (node2 == null) {
			while (node1 != null) {
				tail.setNext(node1);
				node1 = node1.getNext();
			}
		}
		return head;
	}
	
	//使用递归
	public static Node<Integer> merge2(Node<Integer> node1, Node<Integer> node2) {
		Node<Integer> head;
		Node<Integer> oldNode;
		if (node1 == null && node2 == null) {
			throw new IllegalArgumentException("输入的链表不能为空");
		}
		if (node1 != null && node2 != null) {
			if (node1.get() >= node2.get()) {
				oldNode = node2;
				node2 = node2.getNext();
				head = merge2(node1, node2);
			} else {
				oldNode = node1;
				node1 = node1.getNext();
				head = merge2(node1, node2);
			}
			oldNode.setNext(head);
			head = oldNode;
		} else if (node1 == null) {
			while (node2.getNext() != null) {
				head = node2;
				node2 = node2.getNext();
				head.setNext(node2);
			}
			head = node2;
		} else {
			while (node1.getNext() != null) {
				head = node1;
				node2 = node1.getNext();
				head.setNext(node1);
			}
			head = node1;
		}
		return head;
	}

}

File: src/nativeMain/kotlin/com.qiaoyuang.algorithm/round1/Questions25.kt
package com.qiaoyuang.algorithm.round1

fun test25() {
    printlnResult(testCase1())
    printlnResult(testCase2())
    printlnResult(testCase3())
}

/**
 * Questions25: Merge two linked lists be sorted those have benn sorted
 */
private fun <T : Comparable<T>> mergeTwoMergedLinkedList(head1: SingleDirectionNode<T>, head2: SingleDirectionNode<T>): SingleDirectionNode<T> {
    var current1: SingleDirectionNode<T>? = head1
    var current2: SingleDirectionNode<T>? = head2
    val head = if (head1.element <= head2.element) {
        current1 = current1?.next
        head1
    } else {
        current2 = current2?.next
        head2
    }
    var current = head
    while (current1 != null || current2 != null) {
        when {
            current1 == null -> {
                current.next = current2!!
                current = current2
                current2 = current2.next
            }
            current2 == null -> {
                current.next = current1
                current = current1
                current1 = current1.next
            }
            else -> if (current1.element <= current2.element) {
                current.next = current1
                current = current1
                current1 = current1.next
            } else {
                current.next = current2
                current = current2
                current2 = current2.next
            }
        }
    }
    return head
}

private fun testCase1(): Pair<SingleDirectionNode<Int>, SingleDirectionNode<Int>> {
    val node0 = SingleDirectionNode(0)
    val node1 = SingleDirectionNode(1)
    val node2 = SingleDirectionNode(2)
    val node3 = SingleDirectionNode(3)
    val node4 = SingleDirectionNode(4)
    val node5 = SingleDirectionNode(5)
    val node6 = SingleDirectionNode(6)
    val node7 = SingleDirectionNode(7)
    val node8 = SingleDirectionNode(8)
    val node9 = SingleDirectionNode(9)
    node0.next = node2
    node2.next = node4
    node4.next = node6
    node6.next = node8
    node1.next = node3
    node3.next = node5
    node5.next = node7
    node7.next = node9
    return node0 to node1
}

private fun testCase2(): Pair<SingleDirectionNode<Int>, SingleDirectionNode<Int>> {
    val node0 = SingleDirectionNode(0)
    val node1 = SingleDirectionNode(1)
    val node2 = SingleDirectionNode(2)
    val node3 = SingleDirectionNode(3)
    val node4 = SingleDirectionNode(4)
    val node5 = SingleDirectionNode(5)
    val node6 = SingleDirectionNode(6)
    val node7 = SingleDirectionNode(7)
    val node8 = SingleDirectionNode(8)
    val node9 = SingleDirectionNode(9)
    node0.next = node1
    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node5
    node5.next = node6
    node7.next = node8
    node8.next = node9
    return node0 to node7
}

private fun testCase3(): Pair<SingleDirectionNode<Int>, SingleDirectionNode<Int>> {
    val node0 = SingleDirectionNode(0)
    val node1 = SingleDirectionNode(1)
    val node2 = SingleDirectionNode(2)
    val node3 = SingleDirectionNode(3)
    val node4 = SingleDirectionNode(4)
    val node5 = SingleDirectionNode(5)
    val node6 = SingleDirectionNode(6)
    val node7 = SingleDirectionNode(7)
    val node8 = SingleDirectionNode(8)
    val node9 = SingleDirectionNode(9)
    node0.next = node1
    node1.next = node2
    node2.next = node3
    node3.next = node4
    node4.next = node5
    node5.next = node6
    node6.next = node8
    node8.next = node9
    return node7 to node0
}

private fun printlnResult(pair: Pair<SingleDirectionNode<Int>, SingleDirectionNode<Int>>) {
    val (head1, head2) = pair
    print("The linked list 1 is: ")
    printlnLinkedList(head1)
    print("The linked list 2 is: ")
    printlnLinkedList(head2)
    print("After the merger: ")
    val head = mergeTwoMergedLinkedList(head1, head2)
    printlnLinkedList(head)
}